Adjustments to SPEC7 per recent reviews (#3)
stefanv/specs: spec-0007/index.md

We suggest implementing these principles by:

> - deprecating user control of downstream library behavior with `numpy.random.seed`,

Reviewer: I'm not sure what this sentence says now. Perhaps it should be "deprecating libraries dependence on `numpy.random.seed` for global random state"?

Author: I'm not recall whether this was motivated by one of Ralf's comments. I noted at some point that we didn't want to suggest that `np.random.seed` itself is deprecated. Libraries can certainly continue to depend on `np.random.seed` for global random state (e.g. in tests), and users can continue to use `np.random.seed` for whatever they want (independent of the library). What is not allowed anymore is users calling `np.random.seed` to manipulate the random state used within library code.

Reviewer: In that regard, the sentence was quite clear: stop using `np.random.seed` to control random state. I don't think it was saying "`np.random.seed` no longer controls random state".

But, to be explicit, we could say something along the lines of: "In NumPy, `numpy.random.seed` sets a global random seed. Downstream libraries should not make use of this global seed."

Alternatively, we keep the wording as is, since you have already added a sentence later that makes it explicit that `numpy.random.seed` is not deprecated?

Author: ```suggestion
- deprecating the use of `numpy.random.seed` to control the random state,
```

- deprecating the use of `random_state`/`seed` arguments in favor of a consistent `rng` argument, and
- using `numpy.random.default_rng` to validate the `rng` argument and instantiate a `Generator`.

> Note that `numpy.random.default_rng` does not accept instances of `RandomState`, so user control of library behavior with instances of `RandomState` is effectively deprecated, too.

Reviewer: Again, not clear what the "user control of library behavior" distinction is  about; I'll go and look at Ralf's review comments that necessitated this change.

Author: I'm trying to make it explicit that it is an interaction between users and libraries that is no longer allowed. 

I don't think we can just leave the old wording here and add the next change because the previous wording:

> deprecating the use of `numpy.random.seed` to control the random state

and 

> use of `RandomState` to control the seed is effectively deprecated

seem contradicted by the new:

> neither `np.random.seed` nor `np.random.RandomState` themselves are deprecated

After all, what can you do with `np.random.seed` besides control the seed?

The point is that there is not a blanket deprecation of these features; it's something more specific: code inside the library may no longer be controlled by a use of `np.random.seed` outside the library or a `RandomState` passed to a library function.

I will revert these changes but bring up the concern with the old wording in the main repo. Maybe there will be other suggestions, or maybe others will agree with you that it is clear enough as it was.

Author: ```suggestion
Note that `numpy.random.default_rng` does not accept instances of `RandomState`, so use of `RandomState` to control the seed is effectively deprecated, too.
```

That said, neither `np.random.seed` nor `np.random.RandomState` themselves are deprecated, so they may still be used in some contexts (e.g. by developers for generating unit test data).

> ## Motivation

Reviewer: Motivation is quite broad for a SPEC, so I'll have to add a few sentences here. It's certainly broader than adopting Generators.

Author: Great. I mentioned above and in https://github.com/scientific-python/specs/pull/180#discussion_r1705091496 that I would only move this to its own section and would leave the addition of content to you.
